File: algorithms/Arrays/Hard/FourNumberSum/VSDFourNumberSum.py
# Copyrights to vsdevelopers.io
# For more programs visit vsdevelopers.io
# Python program to solve 3 Number Sum


def twosum(arr,target):
    res=[]
    low,high=0,len(arr)-1

    while low<high:
        cursum=arr[low]+arr[high]
        if cursum<target or (low>0 and arr[low]==arr[low-1]):
            low+=1
        elif cursum>target or (high<len(arr)-1 and arr[high]==arr[high+1]):
            high-=1
        else:
            res.append([arr[low],arr[high]])
            low+=1
            high-=1
    return res


def ksum(arr,target,k):
    res=[]
    if not arr:
        return res
    
    avg=target//k

    if avg<arr[0] or avg>arr[-1]:
        return res
    
    if k==2:
        return twosum(arr,target)
    for i in range(len(arr)):
        if i==0 or arr[i-1]!=arr[i]:
            for s in ksum(arr[i+1:],target-arr[i],k-1):
                res.append([arr[i]]+s)
    return res

if __name__=="__main__":
    arr=[1,2,3,4,5,6,7,8,9,10]
    arr.sort()
    target=10
    print(ksum(arr,target,4))

File: algorithms/Arrays/Hard/FourNumberSum/VSDFourNumberSum.java
// Copyrights to venkys.io
// For more programs visit venkys.io 
// Java program for FourNumberSum



import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main{
    static List<List<Integer>> fourSum(int[] arr, int target) {
        ArrayList<List<Integer>> res = new ArrayList<List<Integer>>();
        
        if (arr == null || arr.length == 0)
            return res;
        
        int n = arr.length;
    
        Arrays.sort(arr); 
    
        for (int i = 0; i < n; i++) {
        
            int target_3 = target - arr[i];
        
            for (int j = i + 1; j < n; j++) {
            
                int target_2 = target_3 - arr[j];
            
                int front = j + 1;
                int back = n - 1;
            
                while(front < back) {
                
                    int two_sum = arr[front] + arr[back];
                
                    if (two_sum < target_2) front++;
                
                    else if (two_sum > target_2) back--;
                
                    else {
                    
                        List<Integer> quad = new ArrayList<>(); 
                        quad.add(arr[i]);
                        quad.add(arr[j]);
                        quad.add(arr[front]);
                        quad.add(arr[back]);
                        res.add(quad);
                    
                     
                        while (front < back && arr[front] == quad.get(2)) ++front;
                    
                        
                        while (front < back && arr[back] == quad.get(3)) --back;
                
                    }
                }
                
     
                while(j + 1 < n && arr[j + 1] == arr[j]) ++j;
            }
        
       
            while (i + 1 < n && arr[i + 1] == arr[i]) ++i;
        
        }
        
    
        return res;
    }


	public static void main (String[] args) {
		int arr[] = {1,0,-1,0,-2,2}; 
		int target = 0;

		List<List<Integer>> ls = fourSum(arr, target); 

		for(int i = 0;i<ls.size();i++) {
		    for(int j = 0;j<ls.get(i).size();j++) {
		        System.out.print(ls.get(i).get(j) + " "); 
		    }
		    System.out.println(); 
		}
	}
}